### PR TITLE
[Bridge] Checkpoint gossip

### PR DESCRIPTION
**Pull request by dbrajovic**
# Description

PR that implements checkpoint gossiping. 

Fixes EDGE-458

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [ ] I have added sufficient documentation in code

## Testing

- [ ] I have tested this code with the official test suite
- [ ] I have tested this code manually

### CODE REVIEW

File: bridge/checkpoint/transport/proto/transport.proto [hunk 1]
```diff
@@ -0,0 +1,22 @@
+syntax = "proto3";
+
+package v1;
+
+option go_package = "/bridge/checkpoint/transport/proto";
+
+message SignedMessage {
```
Comment (Kourin1996) [inline bridge/checkpoint/transport/proto/transport.proto]: @dbrajovic I noticed I can't use same name among all protocs in project. Build will fail. Can you rename this to some other name?

**Comment by Kourin1996** [comment]
@dbrajovic Please add build path for checkpoint into `make protoc` and run it.

File: bridge/checkpoint/transport/proto/transport.proto (hunk 1 above)
Comment (dbrajovic) [reply bridge/checkpoint/transport/proto/transport.proto]: What's wrong with having this one ? It's encapsulated within the `checkpoint` module and represents the same concept also used in `statesync`. 

Other than that, I'm open to naming suggestions.

Comment (Kourin1996) [reply bridge/checkpoint/transport/proto/transport.proto]: @dbrajovic For now, you can see this error on building. Please check it

**Review by Kourin1996 — APPROVED** [review]
Looks good to me 👍

**Review by 0xAleksaOpacic — APPROVED** [review]
LGTM

File: bridge/checkpoint/transport/transport.go [hunk 2]
```diff
@@ -62,15 +66,36 @@ func (t *libp2pGossipTransport) Subscribe(handler func(interface{})) error {
 
 		// and call handler
 
-		// protoMessage, ok := obj.(*proto.SignedMessage)
-		// if !ok {
-		// 	t.logger.Warn("received unexpected typed message", "message", obj)
+		protoMsg, ok := obj.(*proto.CheckpointMessage)
+		if !ok {
+			t.logger.Warn("received unexpected typed message", "message", obj)
```
Comment (zivkovicmilos) [inline bridge/checkpoint/transport/transport.go]: Why is this a warning and not an error?

File: bridge/checkpoint/transport/transport.go [hunk 3]
```diff
@@ -62,15 +66,36 @@ func (t *libp2pGossipTransport) Subscribe(handler func(interface{})) error {
 
 		// and call handler
```
Comment (zivkovicmilos) [inline bridge/checkpoint/transport/transport.go]: Are these TODOs leftover?

File: bridge/checkpoint/transport/transport.go [hunk 4]
```diff
@@ -62,15 +66,36 @@ func (t *libp2pGossipTransport) Subscribe(handler func(interface{})) error {
...
+		var message interface{}
+
+		switch protoMsg.Type {
+		case proto.CheckpointMessage_CHECKPOINT:
+			checkpoint := ctypes.Checkpoint{}
+			if err := checkpoint.UnmarshalRLP(protoMsg.Payload); err != nil {
+				t.logger.Error("unable to unmarshal payload from message", "err", err)
+
+				return
+			}
```
Comment (zivkovicmilos) [inline bridge/checkpoint/transport/transport.go]: This unmarshalling can be moved below the switch, so in your case statements you only set the unmarshaller base

**Review by zivkovicmilos — APPROVED** [review]
Looks great 💯 

Very nice touch of using the payload and RLP marshalling it based on the type - it saves us having to define more topics for the clunky pubsub implementation 👏 

Please resolve the linting errors 🙏

File: bridge/checkpoint/transport/transport.go (hunk 4 above)
Comment (zivkovicmilos) [reply bridge/checkpoint/transport/transport.go]: Resolving this as you need to create a concrete struct object after unmarshalling, and I've found no easy and cleaner way to do this

File: bridge/checkpoint/transport/transport.go (hunk 2 above)
Comment (dbrajovic) [reply bridge/checkpoint/transport/transport.go]: [Changed to Error](https://github.com/0xPolygon/polygon-edge/pull/468/commits/d22b06d42c1bb9df130990e51b2785d206ce5c84)

File: bridge/checkpoint/transport/transport.go (hunk 3 above)
Comment (dbrajovic) [reply bridge/checkpoint/transport/transport.go]: [Removed](https://github.com/0xPolygon/polygon-edge/pull/468/commits/bac6047ea80243c7eabe7ad4b8224fe1a4ad058e)

**Pull request merged**